### PR TITLE
[PyTorch] Update PyTorch FSDP2 test to cover all TE layer types

### PR DESCRIPTION
**Pull request by denera**
# Description

This PR expands the FSDP2 test to cover fused and composite layers.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/distributed/run_fsdp2_model.py [hunk 1]
```diff
@@ -148,24 +160,41 @@ def _train(args):
...
     # Apply FSDP/HSDP
-    custom_attrs = save_custom_attrs(model)
-    for sub_module in model.modules():
-        if any(
-            isinstance(sub_module, sub_module_to_wrap) for sub_module_to_wrap in sub_modules_to_wrap
-        ):
-            fully_shard(sub_module, mesh=mesh)
-    fully_shard(model, mesh=mesh)
-    restore_custom_attrs(model, custom_attrs)
+    mp_policy = MixedPrecisionPolicy(
```
Comment (youngeunkwon0405) [inline tests/pytorch/distributed/run_fsdp2_model.py:163]: @denera is there a reason to specify mp_policy instead of using the default?

Comment (denera) [reply tests/pytorch/distributed/run_fsdp2_model.py:163]: > @denera is there a reason to specify mp_policy instead of using the default?

I added this in the process of debugging an FSDP2 bug report. We're finding out that gradient values differ quite wildly between using `torch.amp.autocast` vs `MixedPrecisionPolicy`, and the latter can occasionally even cause NaNs to appear in the backward pass.

Comment (youngeunkwon0405) [reply tests/pytorch/distributed/run_fsdp2_model.py:163]: I see. Thank you for the information.

File: tests/pytorch/distributed/run_fsdp2_model.py [hunk 2]
```diff
@@ -148,24 +160,41 @@ def _train(args):
...
-        target = torch.randn(args.batch_size, args.output_size).to(device)
-        loss = F.mse_loss(output, target)
+        input_data = torch.randn(inp_shape).to(device)
+        target = torch.randn(out_shape).to(device)
+        with torch.autograd.detect_anomaly():
+            with torch.amp.autocast(enabled=not args.fp8_init, device_type="cuda"):
+                with te.fp8_autocast(enabled=args.fp8_init, fp8_recipe=fp8_recipe):
+                    output = model(input_data)
+                    loss = F.mse_loss(output, target)
         loss.backward()
```
Comment (ksivaman) [inline tests/pytorch/distributed/run_fsdp2_model.py:198]:
```suggestion
        with torch.autograd.detect_anomaly():
            with torch.amp.autocast(enabled=not args.fp8_init, device_type="cuda"):
                with te.fp8_autocast(enabled=args.fp8_init, fp8_recipe=fp8_recipe):
                    output = model(input_data)
                    loss = F.mse_loss(output, target)
            loss.backward()
```

Comment (ksivaman) [reply tests/pytorch/distributed/run_fsdp2_model.py:198]: Backward inside context manager as in [example](https://docs.pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly) given in the docs.

**Review by ksivaman — APPROVED** [review]

**Comment by ptrendx** [comment]
The FSDP2 enablement PR #2245 was merged that also includes changes from this PR.
